File: backend/app/schemas/book.py
"""Pydantic schemas for Book API responses."""

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class BookPublic(BaseModel):
    """
    Book data for API responses - sourced from DCS.

    This schema represents books fetched from DCS without local sync.
    The ID is the DCS book ID (integer).
    """

    id: int  # DCS book ID
    name: str  # book_name in DCS
    title: str | None = None  # book_title in DCS
    publisher_id: int
    publisher_name: str
    cover_url: str | None = None
    activity_count: int = 0
    created_at: datetime | None = None
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class BookResponse(BaseModel):
    """
    Book response schema for API.

    DEPRECATED: Use BookPublic instead. This schema is kept for backwards compatibility
    but will be removed in a future version.
    """

    model_config = ConfigDict(from_attributes=True)

    id: int  # Changed from UUID to int (DCS book ID)
    title: str | None = None
    book_name: str | None = None  # DCS book_name
    publisher_name: str
    language: str | None = None
    category: str | None = None
    cover_image_url: str | None = None
    activity_count: int = 0
    dream_storage_id: str | None = None  # For backwards compatibility


class BookListResponse(BaseModel):
    """List of books response schema."""

    items: list[BookResponse]
    total: int
    skip: int = 0
    limit: int = 20


class ActivityResponse(BaseModel):
    """Activity response schema for API."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    book_id: int  # DCS book ID (int) - named book_id for frontend compatibility
    module_name: str | None = None
    page_number: int | None = None
    activity_type: str
    title: str | None = None
    config_json: dict = {}
    order_index: int = 0


class ActivityDetailResponse(ActivityResponse):
    """Detailed activity response with full config."""

    config: dict  # Full config_json


class BookDetailResponse(BookResponse):
    """Detailed book response with activities."""

    activities: list[ActivityResponse]


class BookSyncResponse(BaseModel):
    """Book synchronization response schema."""

    success: bool
    books_synced: int
    books_created: int
    books_updated: int
    activities_created: int
    errors: list[str]
    message: str = "Book sync completed"


# --- Story 8.2: Page-Based Activity Selection ---


class PageInfo(BaseModel):
    """Information about a single page in a book module."""

    page_number: int
    activity_count: int
    thumbnail_url: str


class ModulePages(BaseModel):
    """A module with its pages containing activities."""

    name: str
    pages: list[PageInfo]


class BookPagesResponse(BaseModel):
    """Response for book pages endpoint."""

    book_id: int  # Changed from UUID to int (DCS book ID)
    modules: list[ModulePages]
    total_pages: int
    total_activities: int


class PageActivityResponse(BaseModel):
    """Activity response for page-based selection."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str | None
    activity_type: str
    section_index: int
    order_index: int


# --- Story 8.2 Enhanced: Page Viewer with Activity Markers ---


class ActivityCoords(BaseModel):
    """Coordinates for an activity marker on a page."""

    x: int
    y: int
    w: int
    h: int


class ActivityMarker(BaseModel):
    """Activity marker with position and metadata for page viewer."""

    id: uuid.UUID
    title: str | None
    activity_type: str
    section_index: int
    coords: ActivityCoords | None


class PageDetail(BaseModel):
    """Detailed page information including image and activity markers."""

    page_number: int
    image_url: str
    module_name: str
    activities: list[ActivityMarker]


class ModuleInfo(BaseModel):
    """Module metadata for navigation shortcuts."""

    name: str
    first_page_index: int  # Index in the flat pages array
    page_count: int


class ModulePagesDetail(BaseModel):
    """Module with detailed page information for page viewer."""

    name: str
    pages: list[PageDetail]


class BookPagesDetailResponse(BaseModel):
    """Enhanced book pages response with activity coordinates for page viewer."""

    book_id: int  # Changed from UUID to int (DCS book ID)
    modules: list[ModuleInfo]  # Module shortcuts for navigation
    pages: list[PageDetail]  # Flat list of ALL pages in order
    total_pages: int
    total_activities: int


# --- Story 9.5: Activity Selection Tabs ---


class PageWithActivities(BaseModel):
    """Page information with activity IDs for bulk selection."""

    page_number: int
    thumbnail_url: str
    activity_count: int
    activity_ids: list[uuid.UUID]


class ModuleWithActivities(BaseModel):
    """Module information with pages and activity IDs for bulk selection."""

    name: str
    page_start: int
    page_end: int
    activity_count: int
    activity_ids: list[uuid.UUID]
    pages: list[PageWithActivities]


class BookStructureResponse(BaseModel):
    """Book structure response with modules and pages for activity selection tabs."""

    book_id: int  # Changed from UUID to int (DCS book ID)
    modules: list[ModuleWithActivities]
    total_pages: int
    total_activities: int


# --- Story 10.3: Video Attachment to Assignments ---


class VideoInfo(BaseModel):
    """Video file information from Dream Central Storage."""

    path: str  # Relative path like "videos/chapter1.mp4"
    name: str  # File name like "chapter1.mp4"
    size_bytes: int  # File size
    has_subtitles: bool  # True if .srt file exists


class BookVideosResponse(BaseModel):
    """Response for book videos endpoint."""

    book_id: int  # Changed from UUID to int (DCS book ID)
    videos: list[VideoInfo]
    total_count: int

File: frontend/src/types/book.ts
/**
 * Book and Activity type definitions for Dream LMS.
 * Story 3.6: Book Catalog Browsing for Teachers
 */

/**
 * Activity type enumeration matching backend ActivityType
 */
export type ActivityType =
  | "dragdroppicture"
  | "dragdroppicturegroup"
  | "matchTheWords"
  | "circle"
  | "markwithx"
  | "puzzleFindWords"
  | "fillSentencesWithDots"
  | "fillpicture"

/**
 * Activity types that have implemented players and can be assigned
 * Excludes: fillSentencesWithDots, fillpicture (not yet implemented)
 */
export const SUPPORTED_ACTIVITY_TYPES: ReadonlySet<ActivityType> = new Set([
  "dragdroppicture",
  "dragdroppicturegroup",
  "matchTheWords",
  "circle",
  "markwithx",
  "puzzleFindWords",
])

/**
 * Check if an activity type is supported for assignment
 */
export function isActivityTypeSupported(type: ActivityType): boolean {
  return SUPPORTED_ACTIVITY_TYPES.has(type)
}

/**
 * Book data from API
 *
 * Note: Book IDs are now integers (DCS book IDs) instead of UUIDs.
 * Books are fetched from Dream Central Storage on-demand.
 */
export interface Book {
  id: number // Changed from string to number (DCS book ID)
  dream_storage_id: string
  title: string
  publisher_id: number // DCS publisher ID for logo URL
  publisher_name: string
  description: string | null
  cover_image_url: string | null
  activity_count: number
}

/**
 * Activity data from API
 */
export interface Activity {
  id: string
  book_id: number // Changed from string to number (DCS book ID)
  activity_type: ActivityType
  title: string | null
  config_json: any
  order_index: number
}

/**
 * Paginated book list response
 */
export interface BookListResponse {
  items: Book[]
  total: number
  skip: number
  limit: number
}

/**
 * Filter parameters for book catalog
 */
export interface BooksFilter {
  search?: string
  publisher?: string
  activity_type?: ActivityType
  skip?: number
  limit?: number
}

/**
 * Activity type display configuration
 */
export const ACTIVITY_TYPE_CONFIG: Record<
  ActivityType,
  {
    label: string
    color: string
    badgeVariant: "default" | "secondary" | "destructive" | "outline"
  }
> = {
  dragdroppicture: {
    label: "Drag & Drop",
    color: "blue",
    badgeVariant: "default",
  },
  dragdroppicturegroup: {
    label: "Drag & Drop Group",
    color: "blue",
    badgeVariant: "default",
  },
  matchTheWords: {
    label: "Match Words",
    color: "green",
    badgeVariant: "secondary",
  },
  circle: {
    label: "Circle",
    color: "purple",
    badgeVariant: "outline",
  },
  markwithx: {
    label: "Mark with X",
    color: "orange",
    badgeVariant: "outline",
  },
  puzzleFindWords: {
    label: "Find Words",
    color: "pink",
    badgeVariant: "secondary",
  },
  fillSentencesWithDots: {
    label: "Fill Sentences",
    color: "yellow",
    badgeVariant: "outline",
  },
  fillpicture: {
    label: "Fill Picture",
    color: "teal",
    badgeVariant: "secondary",
  },
}

// --- Story 8.2: Page-Based Activity Selection ---

/**
 * Information about a single page in a book module
 */
export interface PageInfo {
  page_number: number
  activity_count: number
  thumbnail_url: string
}

/**
 * A module with its pages containing activities
 */
export interface ModulePages {
  name: string
  pages: PageInfo[]
}

/**
 * Response from GET /api/v1/books/{book_id}/pages
 */
export interface BookPagesResponse {
  book_id: number // Changed from string to number (DCS book ID)
  modules: ModulePages[]
  total_pages: number
  total_activities: number
}

/**
 * Activity response for page-based selection (simplified)
 */
export interface PageActivity {
  id: string
  title: string | null
  activity_type: ActivityType
  section_index: number
  order_index: number
}

// --- Story 8.2 Enhanced: Page Viewer with Activity Markers ---

/**
 * Coordinates for an activity marker on a page
 */
export interface ActivityCoords {
  x: number
  y: number
  w: number
  h: number
}

/**
 * Activity marker with position and metadata for page viewer
 */
export interface ActivityMarker {
  id: string
  title: string | null
  activity_type: ActivityType
  section_index: number
  coords: ActivityCoords | null
  config: Record<string, unknown>
}

/**
 * Audio marker for page viewer
 */
export interface AudioMarker {
  id: string
  src: string
  x: number
  y: number
  width: number
  height: number
}

/**
 * Video marker for page viewer
 */
export interface VideoMarker {
  id: string
  src: string
  poster: string | null
  subtitle_src: string | null
  x: number
  y: number
  width: number
  height: number
}

/**
 * Fill answer marker for page viewer (clickable answer areas)
 */
export interface FillAnswerMarker {
  id: string
  x: number
  y: number
  width: number
  height: number
  text: string
}

/**
 * Detailed page information including image and activity markers
 */
export interface PageDetail {
  page_number: number
  image_url: string
  module_name: string
  activities: ActivityMarker[]
  audio?: AudioMarker[]
  video?: VideoMarker[]
  fill_answers?: FillAnswerMarker[]
}

/**
 * Module metadata for navigation shortcuts
 */
export interface ModuleInfo {
  name: string
  first_page_index: number // Index in the flat pages array
  page_count: number
}

/**
 * Module with detailed page information for page viewer
 * @deprecated Use ModuleInfo + flat pages list instead
 */
export interface ModulePagesDetail {
  name: string
  pages: PageDetail[]
}

/**
 * Enhanced book pages response with activity coordinates for page viewer
 */
export interface BookPagesDetailResponse {
  book_id: number // Changed from string to number (DCS book ID)
  modules: ModuleInfo[] // Module shortcuts for navigation
  pages: PageDetail[] // Flat list of ALL pages in order
  total_pages: number
  total_activities: number
}

// --- Story 9.5: Activity Selection Tabs ---

/**
 * Page information with activity IDs for bulk selection
 */
export interface PageWithActivities {
  page_number: number
  thumbnail_url: string
  activity_count: number
  activity_ids: string[]
}

/**
 * Module information with pages and activity IDs for bulk selection
 */
export interface ModuleWithActivities {
  name: string
  page_start: number
  page_end: number
  activity_count: number
  activity_ids: string[]
  pages: PageWithActivities[]
}

/**
 * Book structure response with modules and pages for activity selection tabs
 */
export interface BookStructureResponse {
  book_id: number // Changed from string to number (DCS book ID)
  modules: ModuleWithActivities[]
  total_pages: number
  total_activities: number
}
